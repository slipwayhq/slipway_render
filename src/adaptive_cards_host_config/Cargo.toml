[package]
name = "adaptive_cards_host_config"
version.workspace = true
edition.workspace = true
authors.workspace = true

[dependencies]
serde = { workspace = true, features = ["derive"] }

[build-dependencies]
populate_schema_defaults = { path = "../populate_schema_defaults" }
