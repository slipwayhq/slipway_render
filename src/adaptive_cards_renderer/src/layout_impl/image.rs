use std::{cell::RefCell, rc::Rc};

use adaptive_cards::{BlockElementHeight, ImageSize, ImageStyle, StringOrBlockElementHeight};
use image::{RgbaImage, imageops::FilterType};
use imageproc::{
    drawing::{Canvas, draw_filled_rect_mut},
    point::Point,
    rect::Rect,
};
use taffy::{Dimension, Style, TaffyTree};

use crate::{
    ElementLayoutData,
    element_layout_data::ElementTaffyData,
    errors::{RenderError, TaffyErrorToRenderError},
    host_config_utils::StringToColor,
    layout_context::LayoutContext,
    layout_scratch::LayoutScratch,
    masked_image::MaskedImage,
    utils::{ClampToI32, ClampToU32, TaffyLayoutUtils},
};

use super::{
    image_node_context::ImageNodeContext,
    measure::NodeContext,
    utils::{apply_horizontal_alignment, parse_dimension},
};

enum SourceImageSize {
    Fixed,
    Dynamic,
}

impl SourceImageSize {
    fn from_image(image: &adaptive_cards::Image<ElementLayoutData>) -> Self {
        if image.url.contains("$width") || image.url.contains("$height") {
            Self::Dynamic
        } else {
            Self::Fixed
        }
    }
}

impl crate::layoutable::Layoutable for adaptive_cards::Image<ElementLayoutData> {
    fn layout_override(
        &self,
        context: &LayoutContext,
        baseline_style: Style,
        tree: &mut TaffyTree<NodeContext>,
        _scratch: &mut LayoutScratch,
    ) -> Result<ElementTaffyData, RenderError> {
        // Handled by parent
        // self.is_visible
        // self.separator
        // self.spacing
        // self.height

        let mut style = Style { ..baseline_style };
        apply_horizontal_alignment(self.horizontal_alignment, &mut style, context);

        let size = self.size.unwrap_or(ImageSize::Auto);

        if let Some(width) = self.width.as_ref() {
            style.size.width = parse_dimension(width, context)?;
        } else {
            match size {
                adaptive_cards::ImageSize::Auto => {
                    // Image will scale down to fit if needed, but will not scale up to fill the area.
                    style.max_size.width = Dimension::Percent(100.);
                    style.max_size.height = Dimension::Percent(100.);
                }
                adaptive_cards::ImageSize::Stretch => {
                    // Image with both scale down and up to fit as needed.
                    style.max_size.width = Dimension::Percent(100.);
                    style.max_size.height = Dimension::Percent(100.);
                }
                adaptive_cards::ImageSize::Small => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.small as f32);
                }
                adaptive_cards::ImageSize::Medium => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.medium as f32);
                }
                adaptive_cards::ImageSize::Large => {
                    style.size.width =
                        Dimension::Length(context.host_config.image_sizes.large as f32);
                }
            }
        }

        let source_image_size = SourceImageSize::from_image(self);
        match source_image_size {
            SourceImageSize::Dynamic => {
                if let StringOrBlockElementHeight::BlockElementHeight(BlockElementHeight::Auto) =
                    self.height
                {
                    context.host_context.log_warn(&format!(
                        "Image height is set to Auto, but a dynamically sized image is present. This will likely result in a zero height image. Path: {}",
                        context.path
                    ));
                }

                tree.new_leaf(style)
                    .err_context(context)
                    .map(ElementTaffyData::from)
            }
            SourceImageSize::Fixed => {
                let source_image = load_source_image(self, context, None);

                let source_width = source_image.width() as f32;
                let source_height = source_image.height() as f32;

                let max_width = if self.width.as_ref().is_some() {
                    None
                } else {
                    match size {
                        adaptive_cards::ImageSize::Auto => Some(source_width),
                        adaptive_cards::ImageSize::Stretch => None,
                        adaptive_cards::ImageSize::Small => {
                            Some(context.host_config.image_sizes.small as f32)
                        }
                        adaptive_cards::ImageSize::Medium => {
                            Some(context.host_config.image_sizes.medium as f32)
                        }
                        adaptive_cards::ImageSize::Large => {
                            Some(context.host_config.image_sizes.large as f32)
                        }
                    }
                };

                let image_context = ImageNodeContext {
                    source_width,
                    source_height,
                    max_width,
                };

                let node_context = NodeContext::Image(image_context);
                tree.new_leaf_with_context(style, node_context)
                    .err_context(context)
                    .map(ElementTaffyData::from)
            }
        }
    }

    fn draw_override(
        &self,
        context: &LayoutContext,
        tree: &TaffyTree<NodeContext>,
        taffy_data: &ElementTaffyData,
        image: Rc<RefCell<MaskedImage>>,
        _scratch: &mut LayoutScratch,
    ) -> Result<(), RenderError> {
        let node_layout = tree.layout(taffy_data.node_id).err_context(context)?;
        let absolute_rect = node_layout.absolute_rect(context);

        let size = self.size.unwrap_or(ImageSize::Auto);

        let mut source_image = load_source_image(self, context, Some(absolute_rect));

        let (source_image_width, source_image_height) = source_image.dimensions();
        match size {
            ImageSize::Auto => {
                // Image will scale down to fit if needed, but will not scale up to fill the area.
                if source_image_width > absolute_rect.width()
                    || source_image_height > absolute_rect.height()
                {
                    source_image = scale_to_fit_rect(
                        &source_image,
                        absolute_rect.width(),
                        absolute_rect.height(),
                    );
                }
            }
            ImageSize::Stretch => {
                // Image with both scale down and up to fit as needed.
                if source_image_width != absolute_rect.width()
                    && source_image_height != absolute_rect.height()
                {
                    source_image = scale_to_fit_rect(
                        &source_image,
                        absolute_rect.width(),
                        absolute_rect.height(),
                    );
                }
            }
            ImageSize::Small => {
                let target_width = context.host_config.image_sizes.small.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
            ImageSize::Medium => {
                let target_width = context.host_config.image_sizes.medium.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
            ImageSize::Large => {
                let target_width = context.host_config.image_sizes.large.clamp_to_u32();
                scale_to_fit_width(
                    source_image_width,
                    target_width,
                    source_image_height,
                    &mut source_image,
                );
            }
        }

        let horizontal_alignment = self
            .horizontal_alignment
            .unwrap_or(context.inherited.horizontal_alignment);

        let final_position =
            get_new_absolute_position(absolute_rect, &source_image, horizontal_alignment);
        let final_rect = Rect::at(final_position.x, final_position.y)
            .of_size(source_image.width(), source_image.height());

        let style = self.style.unwrap_or(ImageStyle::Default);
        match style {
            ImageStyle::Default => {
                let mut image_mut = image.borrow_mut();
                if let Some(background_color) = self.background_color.as_ref() {
                    let background_color = background_color.to_color()?;
                    draw_filled_rect_mut(&mut *image_mut, final_rect, background_color);
                }
                copy_image(&source_image, &mut image_mut, final_position);
            }
            ImageStyle::Person => {
                // If `style` is set to `person`, apply an ellipse clipping to the image.
                // Assuming the image is square, it'll result in a circle.
                // But if the source image isn't square, it'll result in an ellipse, which is fine.
                let mut image_mut = image.borrow_mut();
                copy_image_within_ellipse(&source_image, &mut image_mut, final_position);
            }
        }

        Ok(())
    }
}

fn load_error_image() -> RgbaImage {
    const SIZE: u32 = 10;
    let mut image = RgbaImage::from_pixel(SIZE, SIZE, image::Rgba([255, 255, 255, 255]));
    for i in 0..SIZE {
        image.put_pixel(i, i, image::Rgba([0, 0, 0, 255]));
        image.put_pixel(i, SIZE - 1 - i, image::Rgba([0, 0, 0, 255]));
    }
    image
}

fn load_source_image(
    image: &adaptive_cards::Image<ElementLayoutData>,
    context: &LayoutContext<'_, '_, '_>,
    absolute_rect: Option<Rect>,
) -> RgbaImage {
    let source_image_size = SourceImageSize::from_image(image);

    let url = match source_image_size {
        SourceImageSize::Dynamic => {
            let Some(absolute_rect) = absolute_rect else {
                panic!("absolute_rect is required when a dynamically sized image is present");
            };

            let mut url = image.url.clone();
            url = url.replace("$width", &absolute_rect.width().to_string());
            url = url.replace("$height", &absolute_rect.height().to_string());
            url
        }
        SourceImageSize::Fixed => image.url.clone(),
    };

    let maybe_image = context
        .host_context
        .load_image_from_url(&url, image.body.as_ref());

    match maybe_image {
        Err(e) => {
            context
                .host_context
                .log_warn(&format!("Failed to load image from URL: {}", url));

            context.host_context.log_warn(&format!(" {}", e.message));

            for inner in e.inner {
                context.host_context.log_warn(&format!(" {}", inner));
            }

            load_error_image()
        }
        Ok(image) => image,
    }
}

fn copy_image(source: &RgbaImage, target: &mut MaskedImage, position: Point<i32>) {
    let x = position.x;
    let y = position.y;
    for sy in 0..source.height() {
        for sx in 0..source.width() {
            let px = source.get_pixel(sx, sy);
            let tx = x + sx.clamp_to_i32();
            let ty = y + sy.clamp_to_i32();
            let Ok(tx) = u32::try_from(tx) else {
                continue;
            };
            let Ok(ty) = u32::try_from(ty) else {
                continue;
            };
            target.draw_pixel(tx, ty, *px);
        }
    }
}

fn copy_image_within_ellipse(source: &RgbaImage, target: &mut MaskedImage, position: Point<i32>) {
    let x = position.x;
    let y = position.y;
    let sw = source.width() as f32;
    let sh = source.height() as f32;
    let cx = sw / 2.0;
    let cy = sh / 2.0;
    let rx = cx;
    let ry = cy;

    for sy in 0..source.height() {
        for sx in 0..source.width() {
            let dx = (sx as f32 - cx) / rx;
            let dy = (sy as f32 - cy) / ry;
            let dist = dx * dx + dy * dy;

            // Dist == 1.0 is on the ellipse boundary, < 1.0 is inside.
            if dist <= 1.0 {
                let px = source.get_pixel(sx, sy);
                let alpha = if dist > 0.9 {
                    // Blend near the edge for a smooth fade.
                    let blend_factor = 1.0 - ((dist - 0.9) / 0.1);
                    (px[3] as f32 * blend_factor) as u8
                } else {
                    px[3]
                };

                let tx = x + sx.clamp_to_i32();
                let ty = y + sy.clamp_to_i32();

                let Ok(tx) = u32::try_from(tx) else {
                    continue;
                };
                let Ok(ty) = u32::try_from(ty) else {
                    continue;
                };

                if tx < target.width() && ty < target.height() {
                    let mut tgt_px = target.get_pixel(tx, ty);
                    let inv_alpha = 255 - alpha;

                    // Simple alpha blending
                    for i in 0..3 {
                        let src_val = px[i] as u16;
                        let tgt_val = tgt_px[i] as u16;
                        let blended = (src_val * alpha as u16 + tgt_val * inv_alpha as u16) / 255;
                        tgt_px[i] = blended as u8;
                    }
                    tgt_px[3] = 255; // Fully opaque result
                    target.draw_pixel(tx, ty, tgt_px);
                }
            }
        }
    }
}

fn get_new_absolute_position(
    absolute_rect: Rect,
    source_image: &RgbaImage,
    horizontal_alignment: adaptive_cards::HorizontalAlignment,
) -> Point<i32> {
    match horizontal_alignment {
        adaptive_cards::HorizontalAlignment::Center => {
            let left = absolute_rect.left()
                + (absolute_rect.width().clamp_to_i32() - source_image.width().clamp_to_i32()) / 2;
            let top = absolute_rect.top()
                + (absolute_rect.height().clamp_to_i32() - source_image.height().clamp_to_i32())
                    / 2;
            Point::new(left, top)
        }
        adaptive_cards::HorizontalAlignment::Left => {
            Point::new(absolute_rect.left(), absolute_rect.top())
        }
        adaptive_cards::HorizontalAlignment::Right => {
            let left = absolute_rect.right() - source_image.width().clamp_to_i32();
            let top = absolute_rect.top();
            Point::new(left, top)
        }
    }
}

fn scale_to_fit_width(
    source_image_width: u32,
    target_width: u32,
    source_image_height: u32,
    source_image: &mut image::ImageBuffer<image::Rgba<u8>, Vec<u8>>,
) {
    if source_image_width != target_width {
        let target_height =
            (source_image_height as f32 * (target_width as f32 / source_image_width as f32)) as u32;
        *source_image = scale_to_fit_rect(&*source_image, target_width, target_height);
    }
}

fn scale_to_fit_rect(img: &RgbaImage, max_width: u32, max_height: u32) -> RgbaImage {
    let (width, height) = img.dimensions();
    let scale = (max_width as f32 / width as f32).min(max_height as f32 / height as f32);
    let new_width = (width as f32 * scale) as u32;
    let new_height = (height as f32 * scale) as u32;
    image::imageops::resize(img, new_width, new_height, FilterType::CatmullRom)
}
