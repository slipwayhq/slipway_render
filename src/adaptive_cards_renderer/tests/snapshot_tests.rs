use std::path::Path;

use adaptive_cards_host_config::HostConfig;
use adaptive_cards_renderer::{
    default_host_config,
    host_context::{HostContext, ResolvedFont},
    render::render_from_str,
    DebugMode,
};
use image::RgbaImage;

const CARD_EXTENSION: &str = ".card.json";

#[test]
fn snapshots() {
    // Get test name from SNAPSHOT_TEST_NAME environment variable.
    let search_path = std::env::var("SNAPSHOT_TEST_NAME")
        .map(|name| format!("snapshot_inputs/{}{}", name, CARD_EXTENSION))
        .unwrap_or_else(|_| format!("snapshot_inputs/*{}", CARD_EXTENSION));

    let debug_mode = match std::env::var("SNAPSHOT_DEBUG_MODE").as_deref() {
        Ok("full") => DebugMode::full(),
        Ok("outlines") => DebugMode::with_outlines(),
        Ok("transparent_masks") => DebugMode::with_transparent_masks(),
        _ => DebugMode::none(),
    };

    insta::with_settings!({}, {
        insta::glob!(&search_path, |path| {
            let path_string = path.to_string_lossy();
            let card_prefix = path_string.strip_suffix(CARD_EXTENSION).unwrap();

            let spec = load_spec_file_for(card_prefix);

            println!("Running snapshot test for {}", card_prefix);

            let json_data = std::fs::read_to_string(path).unwrap();
            let (image, card) = render_from_str(
                &json_data,
                &spec.host_config.unwrap_or_else(default_host_config),
                &MockHostContext {},
                spec.width,
                spec.height,
                debug_mode,
            )
            .unwrap();

            write_image_for(card_prefix, &image);

            insta::assert_json_snapshot!(card);
        });
    });
}

fn write_image_for(card_prefix: &str, image: &RgbaImage) {
    let image_path = format!("{}{}", card_prefix, ".png");
    image.save(image_path).unwrap();
}

fn load_spec_file_for(card_prefix: &str) -> SnapshotTestSpec {
    let spec_file_string = format!("{}{}", card_prefix, ".spec.json");
    let spec_file_path = Path::new(&spec_file_string);
    if spec_file_path.exists() {
        serde_json::from_str::<SnapshotTestSpec>(&std::fs::read_to_string(spec_file_path).unwrap())
            .unwrap()
    } else {
        SnapshotTestSpec {
            width: 1000,
            height: 800,
            host_config: None,
        }
    }
}

#[derive(Debug, serde::Deserialize)]
struct SnapshotTestSpec {
    width: u32,
    height: u32,
    host_config: Option<HostConfig>,
}

// For our tests we're always going to use Roboto, so the results are consistent across platforms.
const ROBOTO_TTF: &[u8] = include_bytes!("../../../fonts/Roboto.ttf");
const ROBOTO_MONO_TTF: &[u8] = include_bytes!("../../../fonts/RobotoMono.ttf");
const AIRPLANE_PNG: &[u8] =
    include_bytes!("../../../src/adaptive_cards_renderer/tests/assets/airplane.png");
struct MockHostContext {}
impl HostContext for MockHostContext {
    fn try_resolve_font(&self, family: &str) -> Option<ResolvedFont> {
        if family.to_lowercase().contains("mono") {
            Some(ResolvedFont {
                family: "Roboto Mono".to_string(),
                data: ROBOTO_MONO_TTF.to_vec(),
            })
        } else {
            Some(ResolvedFont {
                family: "Roboto".to_string(),
                data: ROBOTO_TTF.to_vec(),
            })
        }
    }

    fn run_callout(
        &self,
        _handle: &str,
        _input: &serde_json::Value,
    ) -> Result<RgbaImage, adaptive_cards_renderer::host_context::ComponentError> {
        unimplemented!()
    }

    fn load_image_from_url(
        &self,
        url: &str,
    ) -> Result<RgbaImage, adaptive_cards_renderer::host_context::ComponentError> {
        match url {
            "https://adaptivecards.io/content/airplane.png" => {
                Ok(image::load_from_memory(AIRPLANE_PNG).unwrap().to_rgba8())
            }
            _ => Err(adaptive_cards_renderer::host_context::ComponentError {
                message: format!("Image not found: {}", url),
            }),
        }
    }

    fn warn(&self, message: &str) {
        println!("Warning: {}", message);
    }
}
