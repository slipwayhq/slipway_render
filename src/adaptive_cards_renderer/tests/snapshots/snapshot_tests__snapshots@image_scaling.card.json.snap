---
source: adaptive_cards_renderer/tests/snapshot_tests.rs
expression: snapshot
input_file: adaptive_cards_renderer/tests/snapshot_inputs/image_scaling.card.json
---
{
  "card": {
    "$schema": "../../../../adaptive_cards_data/schema/adaptive-card.schema.json",
    "body": [
      {
        "type": "ColumnSet",
        "columns": [
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "small",
                ".layout": {
                  "rect": {
                    "x": 69,
                    "y": 94,
                    "width": 34,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "small",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 69,
                    "y": 124,
                    "width": 32,
                    "height": 32
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 52,
                "y": 16,
                "width": 68,
                "height": 218
              },
              "placement": "Left"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "medium",
                ".layout": {
                  "rect": {
                    "x": 149,
                    "y": 84,
                    "width": 52,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "medium",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 149,
                    "y": 114,
                    "width": 52,
                    "height": 52
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 132,
                "y": 16,
                "width": 86,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "large",
                ".layout": {
                  "rect": {
                    "x": 247,
                    "y": 60,
                    "width": 100,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "large",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 247,
                    "y": 90,
                    "width": 100,
                    "height": 100
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 230,
                "y": 16,
                "width": 134,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "auto",
                ".layout": {
                  "rect": {
                    "x": 393,
                    "y": 97,
                    "width": 26,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "auto",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 393,
                    "y": 127,
                    "width": 26,
                    "height": 26
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "width": {
              "String": "60px"
            },
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 376,
                "y": 16,
                "width": 60,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "auto",
                ".layout": {
                  "rect": {
                    "x": 465,
                    "y": 78,
                    "width": 116,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "auto",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 465,
                    "y": 108,
                    "width": 116,
                    "height": 64
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "width": {
              "String": "150px"
            },
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 448,
                "y": 16,
                "width": 150,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "stretch",
                ".layout": {
                  "rect": {
                    "x": 627,
                    "y": 52,
                    "width": 116,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "stretch",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 627,
                    "y": 82,
                    "width": 116,
                    "height": 116
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "width": {
              "String": "150px"
            },
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 610,
                "y": 16,
                "width": 150,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "stretch",
                ".layout": {
                  "rect": null,
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "size": "stretch",
                "url": "https://adaptivecards.io/content/airplane.png",
                ".layout": {
                  "rect": {
                    "x": 789,
                    "y": 1,
                    "width": 266,
                    "height": 266
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "width": {
              "String": "300px"
            },
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 772,
                "y": 16,
                "width": 300,
                "height": 218
              },
              "placement": "WithinHorizontal"
            }
          },
          {
            "items": [
              {
                "type": "TextBlock",
                "text": "20px",
                ".layout": {
                  "rect": {
                    "x": 1101,
                    "y": 100,
                    "width": 30,
                    "height": 18
                  },
                  "placement": "Top"
                }
              },
              {
                "type": "Image",
                "url": "https://adaptivecards.io/content/airplane.png",
                "width": "20px",
                ".layout": {
                  "rect": {
                    "x": 1101,
                    "y": 130,
                    "width": 20,
                    "height": 20
                  },
                  "placement": "Bottom"
                }
              }
            ],
            "style": "accent",
            "type": "Column",
            ".layout": {
              "rect": {
                "x": 1084,
                "y": 16,
                "width": 64,
                "height": 218
              },
              "placement": "Right"
            }
          }
        ],
        "height": "stretch",
        ".layout": {
          "rect": {
            "x": 16,
            "y": 16,
            "width": 1168,
            "height": 218
          },
          "placement": "SoleVertical"
        }
      }
    ],
    "horizontalAlignment": "center",
    "version": "1.5",
    "verticalContentAlignment": "center",
    "type": "AdaptiveCard",
    ".layout": {
      "rect": {
        "x": 0,
        "y": 0,
        "width": 1200,
        "height": 250
      }
    }
  },
  "debug_lines": [
    "TREE",
    "└──  FLEX COL [x: 0    y: 0    w: 1200 h: 250  content_w: 1200 content_h: 300  border: l:0 r:0 t:0 b:0, padding: l:16 r:16 t:16 b:16] (NodeId(4294967337))",
    "    └──  FLEX ROW [x: 16   y: 16   w: 1168 h: 218  content_w: 1132 content_h: 268  border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967336))",
    "        ├──  FLEX COL [x: 36   y: 0    w: 68   h: 218  content_w: 68   content_h: 157  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967300))",
    "        │   ├──  LEAF [x: 17   y: 78   w: 34   h: 18   content_w: 34   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967297))",
    "        │   ├──  LEAF [x: 17   y: 96   w: 34   h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967298))",
    "        │   └──  LEAF [x: 17   y: 108  w: 32   h: 32   content_w: 32   content_h: 32   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967299))",
    "        ├──  LEAF [x: 104  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967301))",
    "        ├──  FLEX COL [x: 116  y: 0    w: 86   h: 218  content_w: 86   content_h: 167  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967305))",
    "        │   ├──  LEAF [x: 17   y: 68   w: 52   h: 18   content_w: 51   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967302))",
    "        │   ├──  LEAF [x: 17   y: 86   w: 52   h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967303))",
    "        │   └──  LEAF [x: 17   y: 98   w: 52   h: 52   content_w: 52   content_h: 52   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967304))",
    "        ├──  LEAF [x: 202  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967306))",
    "        ├──  FLEX COL [x: 214  y: 0    w: 134  h: 218  content_w: 134  content_h: 191  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967310))",
    "        │   ├──  LEAF [x: 17   y: 44   w: 100  h: 18   content_w: 31   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967307))",
    "        │   ├──  LEAF [x: 17   y: 62   w: 100  h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967308))",
    "        │   └──  LEAF [x: 17   y: 74   w: 100  h: 100  content_w: 100  content_h: 100  border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967309))",
    "        ├──  LEAF [x: 348  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967311))",
    "        ├──  FLEX COL [x: 360  y: 0    w: 60   h: 218  content_w: 98   content_h: 192  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967315))",
    "        │   ├──  LEAF [x: 17   y: 81   w: 26   h: 18   content_w: 28   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967312))",
    "        │   ├──  LEAF [x: 17   y: 99   w: 26   h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967313))",
    "        │   └──  LEAF [x: 17   y: 111  w: 26   h: 26   content_w: 64   content_h: 64   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967314))",
    "        ├──  LEAF [x: 420  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967316))",
    "        ├──  FLEX COL [x: 432  y: 0    w: 150  h: 218  content_w: 150  content_h: 173  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967320))",
    "        │   ├──  LEAF [x: 17   y: 62   w: 116  h: 18   content_w: 28   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967317))",
    "        │   ├──  LEAF [x: 17   y: 80   w: 116  h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967318))",
    "        │   └──  LEAF [x: 17   y: 92   w: 116  h: 64   content_w: 64   content_h: 64   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967319))",
    "        ├──  LEAF [x: 582  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967321))",
    "        ├──  FLEX COL [x: 594  y: 0    w: 150  h: 218  content_w: 150  content_h: 199  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967325))",
    "        │   ├──  LEAF [x: 17   y: 36   w: 116  h: 18   content_w: 43   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967322))",
    "        │   ├──  LEAF [x: 17   y: 54   w: 116  h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967323))",
    "        │   └──  LEAF [x: 17   y: 66   w: 116  h: 116  content_w: 116  content_h: 116  border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967324))",
    "        ├──  LEAF [x: 744  y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967326))",
    "        ├──  FLEX COL [x: 756  y: 0    w: 300  h: 218  content_w: 300  content_h: 268  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967330))",
    "        │   ├──  LEAF [x: 17   y: -33  w: 266  h: 18   content_w: 43   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967327))",
    "        │   ├──  LEAF [x: 17   y: -15  w: 266  h: 0    content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967328))",
    "        │   └──  LEAF [x: 17   y: -15  w: 266  h: 266  content_w: 266  content_h: 266  border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967329))",
    "        ├──  LEAF [x: 1056 y: 0    w: 12   h: 218  content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967331))",
    "        └──  FLEX COL [x: 1068 y: 0    w: 64   h: 218  content_w: 64   content_h: 151  border: l:0 r:0 t:0 b:0, padding: l:17 r:17 t:17 b:17] (NodeId(4294967335))",
    "            ├──  LEAF [x: 17   y: 84   w: 30   h: 18   content_w: 30   content_h: 18   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967332))",
    "            ├──  LEAF [x: 17   y: 102  w: 30   h: 12   content_w: 0    content_h: 0    border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967333))",
    "            └──  LEAF [x: 17   y: 114  w: 20   h: 20   content_w: 20   content_h: 20   border: l:0 r:0 t:0 b:0, padding: l:0 r:0 t:0 b:0] (NodeId(4294967334))"
  ]
}
